src/Microsoft.ML.FastTree/FastTree.cs

```diff
@@ -3158,7 +3158,14 @@ public void SaveSummary(TextWriter writer, RoleMappedSchema schema)
                 var name = names.GetItemOrDefault(pair.Key).ToString();
                 if (string.IsNullOrEmpty(name))
                     name = $"f{pair.Key}";
-                yield return new KeyValuePair<string, Double>(name, Math.Sqrt(pair.Value) * normFactor);
+                
+                //REVIEW: For categorical split points some gain values can be negative because when gain
+                //map is built we have to reverse engineer what each feature-value's gain was from the total 
+                //gain of split at that node. The reason we chose to do this way was to speed things up and 
+                //reduce memory footprint during training time. A better solution would be to keep track 
+                //of each split point's contribution during training but that is an expensive approach.
+                double value = pair.Value < 0 ? -1 * Math.Sqrt(Math.Abs(pair.Value)) : Math.Sqrt(pair.Value);
```

Comment:
>pair.Value < 0 [](start = 31, length = 14)

I'm not sure I get this, so maybe you can help me out. I think you're encoding some information in the sign. Could we explain this?

Reply:
Split point that have zero document count will have negative gain value instead of zero because we distribute the gain residual evenly across split points when computing individual split point contribution towards the total gain from a node split. 

NodeGain = (bestShiftedGain - gainShift) * trust - usePenalty

We separate out gainshift, trust and usePenalty from the total gain and evenly distribute it across split points. 

One way to avoid negative gain would be set gain value for split points that have zero documents to zero and only distribute residual to non-zero document count split points.


---
In reply to: [194045182](https://github.com/dotnet/machinelearning/pull/320#discussion_r194045182) [](ancestors = 194045182)